libraries/AP_InertialSensor/BatchSampler.cpp

```diff
@@ -0,0 +1,198 @@
+    // @User: Advanced
+    // @Increment: 32
+    AP_GROUPINFO("BAT_CNT",  1, AP_InertialSensor::BatchSampler, _required_count,   1024),
+
+    // @Param: BAT_MASK
+    // @DisplayName: Sensor Bitmask
+    // @Description: Bitmap of which IMUs to log batch data for
+    // @User: Advanced
+    // @Values: 0:None,1:First IMU,255:All
+    // @Bitmask: 0:IMU1,1:IMU2,2:IMU3
+    AP_GROUPINFO("BAT_MASK",  2, AP_InertialSensor::BatchSampler, _sensor_mask,   DEFAULT_IMU_LOG_BAT_MASK),
+
+    AP_GROUPEND
+};
+
+
+extern const AP_HAL::HAL& hal;
+void AP_InertialSensor::BatchSampler::init()
+{
+    if (_sensor_mask == 0) {
+        return;
+    }
+    if (_required_count <= 0) {
+        return;
+    }
+
+    _required_count -= _required_count % 32; // round down to nearest multiple of 32
+
+    const uint32_t total_allocation = 3*_required_count*sizeof(uint16_t);
+    gcs().send_text(MAV_SEVERITY_WARNING, "INS: alloc %u bytes for ISB (free=%u)", total_allocation, hal.util->available_memory());
+
+    data_x = (int16_t*)calloc(_required_count, sizeof(int16_t));
+    data_y = (int16_t*)calloc(_required_count, sizeof(int16_t));
+    data_z = (int16_t*)calloc(_required_count, sizeof(int16_t));
+    if (data_x == nullptr || data_y == nullptr || data_z == nullptr) {
+        free(data_x);
+        free(data_y);
+        free(data_z);
+        data_x = nullptr;
+        data_y = nullptr;
+        data_z = nullptr;
+        gcs().send_text(MAV_SEVERITY_WARNING, "Failed to allocate %u bytes for IMU batch sampling", total_allocation);
+        return;
+    }
+
+    initialised = true;
+}
+
+void AP_InertialSensor::BatchSampler::periodic()
+{
+    if (_sensor_mask == 0) {
+        return;
+    }
+    push_data_to_log();
+}
+
+void AP_InertialSensor::BatchSampler::rotate_to_next_sensor()
+{
+    const uint8_t old_instance = instance;
+    do {
+        instance++;
+        switch(type) {
+        case IMU_SENSOR_TYPE_ACCEL:
+            if (instance >= _imu._accel_count) {
+                type = IMU_SENSOR_TYPE_GYRO;
+                instance = 0;
+                multiplier = multiplier_accel;
+            }
+            break;
+        case IMU_SENSOR_TYPE_GYRO:
+            if (instance >= _imu._gyro_count) {
+                type = IMU_SENSOR_TYPE_ACCEL;
+                instance = 0;
+                multiplier = multiplier_gyro;
+            }
+            break;
+        }
+        if (_sensor_mask & (1U<<instance)) {
+            break;
+        }
+    } while (instance != old_instance);
+}
+
+void AP_InertialSensor::BatchSampler::push_data_to_log()
+{
+    if (!initialised) {
+        return;
+    }
+    if (_sensor_mask == 0) {
+        return;
+    }
+    if (data_write_offset - data_read_offset < samples_per_msg) {
+        // insuffucient data to pack a packet
+        return;
+    }
+    const uint64_t now = AP_HAL::micros64();
+    if (last_sent_ms - now < push_interval_ms) {
```

Comment:
it is not safe to subtract uint64_t and uint32_t - both must be the same width

Reply:
also need to be the same units! don't compare milliseconds and microseconds

Reply:
Will fix.